### PR TITLE
use a custom program streaming format

### PR DESCRIPTION
**Pull request by cuddlefishie**
As part of #291 , this format will reduce the amount of data that needs to be transferred to a bootloader, making iteration times shorter.

An ELF container encodes a lot more information than a simple bootloader needs. Additionally, none of the `#![no_std]` ELF readers support using a streaming format. That means the entire ELF needs to be loaded and held in memory, to then be parsed and then the contents are written out to actual memory locations, essentially using double the memory requirements on the system.

The data that is *kept* from the ELF is essentially:
- entry point
- number of segments to load
- address of segment
- whether segment contains instructions or data
- data for segment
- needed zero padding of a segment

This format allows a bootloader to keep very minimal overhead for the receiving of the data and enables reading the data directly into the memory addresses they should end up in anyway!

It's only a rough estimation, but with this, the required memory for loading a program from some `impl Iterator<Item = u8>` (that is not counted here) should be around 32 bytes!

Because the streaming format strips away information from the ELFs that are not needed, they are strictly smaller in size. (The `all_streams_are_smaller_than_elfs` property test also checks this)

TODOs:
- [x] write tests

### CODE REVIEW

File: bittide/src/Bittide/ProcessingElement/ProgramStream.hs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+--
+-- SPDX-License-Identifier: Apache-2.0
+
+module Bittide.ProcessingElement.ProgramStream (
+    elfStreamStructure,
+    elfStream,
+    ZeroPaddingLength,
+    IsExecutable,
+    Segment,
+    Address
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/ProcessingElement/ProgramStream.hs:11]:
```suggestion
    , elfStreamStructure
    , elfStream
    . ZeroPaddingLength
    , IsExecutable
    , Segment
    , Address
```

File: bittide/src/Bittide/ProcessingElement/ProgramStream.hs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    | PF_R `elem` elfSegmentFlags seg
+    = let
+        segData = elfSegmentData seg
+        fileSz = fromIntegral $ BS.length segData
+        memSz = fromIntegral $ elfSegmentMemSize seg
+        zeroPadding = memSz - fileSz
+        addr = fromIntegral $ elfSegmentPhysAddr seg
+        segEntry   = (False, addr, bytes segData, zeroPadding)
+      in segEntry : acc
```
Comment (martijnbastiaan) [inline bittide/src/Bittide/ProcessingElement/ProgramStream.hs:81]: Deduplicate this with a `goSegment`?

File: firmware/bittide-sys/tests/program_stream.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+lazy_static::lazy_static! {
+    static ref PROG_STREAM_EXEC: PathBuf = program_stream_exec_path().unwrap();
+}
+
+#[test]
+fn find_cabal_stuff() {
+    dbg!(&*PROG_STREAM_EXEC);
+}
+
+unsafe fn write_program_to_memory(input: &mut impl Iterator<Item = u8>) -> Option<()> {
```
Comment (martijnbastiaan) [inline firmware/bittide-sys/tests/program_stream.rs:91]: This seems to always return a trivial `Some()`. What's up with this return type? Shouldn't it return simply nothing?

**Review by martijnbastiaan — CHANGES_REQUESTED** [review]
Could you:

 * Add Haddock documentation to your Haskell functions
 * Write a PR cover letter: it should answer the question why this reduce the amount of memory we need to transfer binaries

But as far as I can tell: looks good!

File: firmware/bittide-sys/tests/program_stream.rs (hunk 3 above)
Comment (martijnbastiaan) [reply firmware/bittide-sys/tests/program_stream.rs:91]: Great makes sense

**Comment by cuddlefishie** [comment]
Added haddock comments, a new property test to make sure the streaming format will always be smaller than an ELF (just a sanity check) and the PR description was updated!

**Review by martijnbastiaan — APPROVED** [review]

**Pull request merged**